add page Database migrations (#889)
strapi/documentation: docs/developer-docs/latest/developer-resources/database-migrations.md

---
title: Database migrations - Strapi Developer Docs
description: Strapi database migrations are ways to modify the database
canonicalUrl: https://docs.strapi.io/developer-docs/latest/development/backend-customization/database-migrations.html
---

# Database migrations (experimental feature)

> Database migrations exist to run one-time queries against the database, typically to modify the tables structure or the data when upgrading the Strapi app.

Reviewer: ```suggestion
Database migrations exist to run one-time queries against the database, typically to modify the tables structure or the data when upgrading the Strapi application.
```

::: callout 🚧  Experimental
The feature is to be considered as unfinished and experimental.
:::

## General logic

> Migration are run using javascript migration files stored in `database/migrations`.

Reviewer: ```suggestion
Migrations are run using JavaScript migration files stored in `./database/migrations`.
```

> Migration files should export 2 functions `up()` and `down()`.

Reviewer: ```suggestion
Migration files should export 2 functions `up()` and `down()`:
```

Reviewer 2: Not sure what these symbols are but probably best to not include them
![image](https://user-images.githubusercontent.com/8593673/167686872-2080c429-5a3c-4510-9dd6-d1b82393ddc7.png)

Reviewer: Yeah, not sure what happened, had to remove some in other places yesterday and my commit was supposed to remove them here, not _add_ them… 😂   🤷 
Thanks for the heads up, Derrick!

> - The `up()`function is used when upgrading (ex: adding a new table `my_new_table`)

Reviewer: ```suggestion
- The `up()`function is used when upgrading (e.g. adding a new table `my_new_table`).
```

- The `down()`function is used to reverse the `up()` function when downgrading (ex: deleting `my_new_table`)

> Strapi will automatically detect migration files and run then, once, at the next startup, in alphabetical order. Every new file will be executed once.

Reviewer: ```suggestion
Strapi will automatically detect migration files and run then, once, at the next startup, in alphabetical order. Every new file is executed once.
```

Migrations are run before the database tables are synced with the content-types schemas.

There is no CLI to run the migration by hand for the moment.

There are no ways to use the downgrade feature for the moment, hence, the `down()`is not usable for the moment.

`up()` and `down()` functions run in a database transaction which means if a query fails during the migration, the whole migration will be cancelled and no changes will be applied to the database. It also means that if an other transaction is created within the migration function, it will be a nested transaction.

## Creating a migration file

> 1. Create a new file in `database/migrations` with a name containing the date and the name of the migration (ex: `2022.05.10T00.00.00.name-of-my-migration.js`). The file name is important because the alphabetical order of the files defines the order in which the migrations have to run.

Reviewer: ```suggestion
To create a migration file: 

1. In the `./database/migrations` folder, create a new file named after the date and the name of the migration (e.g. `2022.05.10T00.00.00.name-of-my-migration.js`). Make sure that the file name follows this naming pattern, because the alphabetical order of the files defines the order in which the migrations have to run.
```

> 2. Fill the migration file with the following template:

Reviewer: ```suggestion
2. Copy and paste the following template in the previously created file:
```

```jsx
'use strict'

async function up(trx) {}

async function down(trx) {}

module.exports = { up, down };
```

> 3. Write the migration code inside the `up()` and `down()` functions.

Reviewer: ```suggestion
3. Fill in the template by adding actual migration code inside the `up()` and `down()` functions.
```

Strapi will automatically detect migration files and run them once at the next startup in alphabetical order. Every new file is executed once.

Migrations are run before the database tables are synced with the content-types schemas.

There is no CLI to run the migration manually for the moment.

There are no ways to use the downgrade feature for the moment, hence, the `down()`is not usable for the moment.

`up()` and `down()` functions run in a database transaction which means if a query fails during the migration, the whole migration will be cancelled and no changes will be applied to the database. It also means that if an other transaction is created within the migration function, it will be a nested transaction.

## Creating a migration file

To create a migration file: 

1. In the `./database/migrations` folder, create a new file named after the date and the name of the migration (e.g. `2022.05.10T00.00.00.name-of-my-migration.js`). Make sure that the file name follows this naming pattern, because the alphabetical order of the files defines the order in which the migrations have to run.

2. Copy and paste the following template in the previously created file:

> ```jsx

Reviewer 2: I'd recommend we offer a more complete example, and I believe we are passing in knex in it's entirety not just transactions.

See this example from the repo I've been referencing about this until we added the docs for it: https://github.com/derrickmehaffy/strapi-example-v4-custom-db/blob/main/database/migrations/test.js

This also doesn't cover the fact you can add .sql files here as well such as in my example here: https://github.com/derrickmehaffy/strapi-example-v4-custom-db/blob/main/database/migrations/testSQL.sql.example

Author: Thank you for the great example!
And yes it's a knex instance but in a transaction mode

'use strict'

async function up(trx) {}

async function down(trx) {}

module.exports = { up, down };
```

3. Fill in the template by adding actual migration code inside the `up()` and `down()` functions.

> `up()` and `down()` receive a [Knex instance](https://knexjs.org/) that can be used to run the database queries.

Reviewer 2: more odd symbols int he functions

::: callout 🚧  Experimental feature
Database migrations are experimental. This feature is still a work in progress and will continue to be updated and improved. In the meantime, feel free to ask for help on the [forum](https://forum.strapi.io/) or on the community [Discord](https://discord.strapi.io).
:::

## Understanding database migration files

Migrations are run using JavaScript migration files stored in `./database/migrations`.

Strapi automatically detects migration files and run them once at the next startup in alphabetical order. Every new file is executed once. Migrations are run before the database tables are synced with the content-types schemas.

::: warning
Currently Strapi does not support down migrations. This means that if you need to revert a migration, you will have to do it manually. It is planned to implement down migrations in the future but no timeline is currently available.
:::

> Migration files should export the function `up()`:

Reviewer: ```suggestion
Migration files should export the function `up()`, which is used when upgrading (e.g. adding a new table `my_new_table`).
```

> - The `up()` function is used when upgrading (e.g. adding a new table `my_new_table`).

Reviewer: ```suggestion
```

## Creating a migration file

To create a migration file:

1. In the `./database/migrations` folder, create a new file named after the date and the name of the migration (e.g. `2022.05.10T00.00.00.name-of-my-migration.js`). Make sure that the file name follows this naming pattern, because the alphabetical order of the files defines the order in which the migrations have to run.

2. Copy and paste the following template in the previously created file:

```jsx
'use strict'

async function up(knex) {}

async function down(knex) {}

module.exports = { up, down };
```

3. Fill in the template by adding actual migration code inside the `up()` and `down()` functions.
`up()` and `down()` receive a [Knex instance](https://knexjs.org/), already in a transaction state, that can be used to run the database queries.

::: details Example of migration file

```jsx
// path: ./database/migrations/2022.05.10T00.00.00.name-of-my-migration.js

module.exports = {
  async up(knex) {
    // You have full access to the Knex.js API with an already initialized connection to the database
    // ------------------------
    // EXAMPLE: renaming a table

>     // knex.schema.renameTable('oldName', 'newName');

Reviewer 2: Should we add await here also?

    // ------------------------
    // EXAMPLE: renaming a column

>     // knex.schema.table('someTable', table => {

Reviewer 2: Should we add await here also?
